docs/development/ADRs/0017-Toolchain-Configuration.md

```diff
@@ -0,0 +1,101 @@
+
+### Environment variables are the primary end user interface
+
+Each user configurable option must be loaded from an environment variable if it is set. If there is an in-code default value it must be overridden by the environment variable.
+
+## Consequences
+
+### Testability is limited
+
+Since the process of loading user configuration is in module scope, it is not repeatable for testing purposes. Module unloading and reloading is too much work and fragile.
+
+Monkey patching can be used to test whether the values in the config module are reflected in newly built toolchains, but this very process sidesteps the actual reading of user configuration, which remains inaccessible to testing.
+
+### Implementation is kept simple
+
+All the logic for how to interpret the user configuration is manually encoded in the high level toolchain building interface. This does not currently require much code. No infrastructure is needed to automatically gather configuration options from toolchain definitions and present them to the user, as the exposed options are hand-picked.
+
+Config variables can be assumed to be fixed for the entirety of a run of any program that uses GT4Py.
+
+### It is not possible to track where configuration values come from.
+
+- the user-set environment variables must override the config module fallback value (can not be tested)
+- the config module value must override the default values of the high-level toolchain building interface (testing up to toolchain implementer, requires monkey patching)
+- the high-level toolchain building interface's defaults should override defaults in the toolchain modules (possibly testable, up to toolchain implementer)
+- arguments passed to the high-level toolchain building interface should override all others, including user config (testing up to toolchain implementer)
+- toolchain instances created without the high-level interface can do whatever they want (not testable in general)
+
+The situation may arise that toolchain instances are used (possibly for good reason) in a program / library using GT4Py, which do not respect user configuration. It remains up to the implementer of such a program or library to communicate this to the user.
+
+## Alternatives Considered
+
+### Warn the user of workflow instances that disregard user configuration
+
+The user would receive a warning, either
+
+- when a toolchain instance is created, which disregards user config, or
+- when such a toolchain is used
+
+The latter was never tried due to the obvious run time overhead in checking before every use. The former turned out not to be very useful for two reasons:
+
+- such a toolchain may have been created but not used to do work for the end user
+- without proper tracking of where configuration comes from, false positives as well as false negatives could not be eliminated.
+
+Implementing tracking was briefly considered but looked like it would be too heavy weight to justify the maintenance burden.
```

Comment:
This paragraph doesn't really sum up the experiment we did. Also where is the
```
class MyStep:
    config_flag = True

os.environ["GT4PY_MYSTEP_CONFIG_FLAG"] = True

my_backend = MyStepFactory(mystep_config_flag=False)
```
part or the sketch of the merging algorithm?

Reply:
The merging algorithm I proposed has traceability:
```
env_defined_vars: dict[str, Any]
config_file_defined_vars: dict[str, Any]
code_defined_vars: dict[str, Any]

var_sources = {
  "env": env_defined_vars,
  "config": config_file_defined_vars,
  "in-code": code_defined_vars
}

effects_per_var_source: dict[str, dict[str, list[str]]] = []

for var_source, vars in var_sources.items():
  res = factory(**vars)
  # compute what attributes are different in res from default, i.e. factory()
  effect = ...
  effects_per_var_source[var_source][var] = effect

for (source_a, effect_a), (source_b, effect_b) in itertools.product(effects_per_var_source.items(), effects_per_var_source.items())
  for name_outer, effect_outer in user_defined_vars_effects.items():
    for name_inner, effect_inner in code_defined_vars_effects.items():
      if intersection(effect_outer, effect_inner):
        print("{source_a}:{name_outer} conflicts with {source_b}:{name_inner}")
```
It prints both source of the conflict and can straightforward be extended to give even more (e.g. line numbers if the sources are from config files). But my main point is I don't see this paragraph to reflect what we concluded the experiment with:

> Algorithmically, warnings are generally possible, but not with FactoryBoy for Defaults. Extensions to multiple SOT are generally also possible and in the simple version also feasible, but whether that is still satisfactory is questionable (because other projects also do not do that at all).

Reply:
Ah, I will add another paragraph about black box analysis of the effects of different configuration sources and clarify that this one is about **tracking** where which value comes from.

Reply:
Also, I don't undestand why you would mention `factory-boy`. The black-box analysis is agnostic, it can always only easily track to the highest level of toolchain building logic. As soon as the user configuration does not apply directly anymore (but only indirectly via toolchain building logic), it would involve replicating some of that logic, or tracking (as in annotating any configuration with it's source and keeping and updating that annotation at every level of logic).